autoAdjustHwloc.hpp

```diff
@@ -23,154 +24,151 @@ class autoAdjust
-		if (depth == HWLOC_TYPE_DEPTH_UNKNOWN)
+		try
 		{
-			printf("*** The number of sockets is unknown\n");
-		}
+			std::vector<hwloc_obj_t> tlcs;
+			tlcs.reserve(16);
+			results.reserve(16);
 
-		for (int i = 0; i < hwloc_get_nbobjs_by_depth(topology, depth); i++)
-		{
-			socket = hwloc_get_obj_by_depth(topology, depth, i);
+			findChildrenByType(hwloc_get_root_obj(topology), HWLOC_OBJ_CACHE,
+				[&tlcs](hwloc_obj_t found) { tlcs.emplace_back(found); } );
+
+			if(tlcs.size() == 0)
+				throw(std::runtime_error("The CPU doesn't seem to have a cache."));
 
-			// search cacheprinter::inst()->print_str("\n**************** Copy&Paste ****************\n\n");
-			for (int j = 0; j < socket->arity; j++)
+			for(hwloc_obj_t obj : tlcs)
+				proccessTopLevelCache(obj);
+
+			printer::inst()->print_str("\n**************** Copy&Paste BEGIN ****************\n\n");
+			printer::inst()->print_str("\"cpu_threads_conf\" :\n[\n");
+
+			for(uint32_t id : results)
 			{
-				hwloc_obj_t nextLvl = socket->children[j];
-				findCache(topology, nextLvl);
+				char str[128];
+				snprintf(str, sizeof(str), "    { \"low_power_mode\" : %s, \"no_prefetch\" : true, \"affine_to_cpu\" : %u },\n",
+					(id & 0x8000000) != 0 ? "true" : "false", id & 0x7FFFFFF);
+				printer::inst()->print_str(str);
 			}
+
+			printer::inst()->print_str("],\n\n**************** Copy&Paste END ****************\n");
+		}
+		catch(const std::runtime_error& err)
+		{
+			printer::inst()->print_msg(L0, "Autoconf FAILED: %s", err.what());
+			printer::inst()->print_str("\nPrinting config for a single thread. Please try to add new ones until the hashrate slows down.\n");
+			printer::inst()->print_str("\n**************** FAILURE Copy&Paste BEGIN ****************\n\n");
+			printer::inst()->print_str("\"cpu_threads_conf\" :\n[\n");
+			printer::inst()->print_str("    { \"low_power_mode\" : false, \"no_prefetch\" : true, \"affine_to_cpu\" : false },\n");
+			printer::inst()->print_str("],\n\n**************** FAILURE Copy&Paste END ****************\n");
 		}
 
 		/* Destroy topology object. */
 		hwloc_topology_destroy(topology);
-		
-		printer::inst()->print_str("],\n\n**************** Copy&Paste END ****************\n");
 	}
 
 private:
+	static constexpr size_t hashSize = 2 * 1024 * 1024;
+	std::vector<uint32_t> results;
 
-	inline void getConfig(hwloc_topology_t topology, hwloc_obj_t obj, size_t& numHashes, int& numCachesLeft)
+	template<typename func>
+	inline void findChildrenByType(hwloc_obj_t obj, hwloc_obj_type_t type, func lambda)
 	{
-		if (obj->type == HWLOC_OBJ_CORE)
+		for(size_t i=0; i < obj->arity; i++)
 		{
-			if (obj)
-			{
-				hwloc_cpuset_t cpuset;
-				/* Get a copy of its cpuset that we may modify. */
-				cpuset = hwloc_bitmap_dup(obj->cpuset);
-				size_t allcpu = hwloc_bitmap_to_ulong(cpuset);
-				/* Get only one logical processor (in case the core is
-				   SMT/hyperthreaded). */
-				hwloc_bitmap_singlify(cpuset);
+			if(obj->children[i]->type == type)
```

Comment:
Quick compile check spat out that `hwloc_obj_type_is_cache()` is undefined, so it looks like 2.X api is a clean break with 1.X that we have currently. It will be fairly trivial to modify once we actually switch over though.